### PR TITLE
[3366]-deploy-find-ftt-to-paas-with-terraform

### PR DESCRIPTION
**Pull request by karthikdfe**
### Context

Deploying the find webapp and point it to the QA API running on Azure.
Deploy via docker (not via buildpack) as it's DfE strategic direction and our apps are already packaged as docker images. The latest docker image published to dockerhub may be used.

### Changes proposed in this pull request

The terraform files should reside in the find repository. We should be able to:
- deploy with the same image: should be no-op
- deploy with a new image: should update the app
- deploy with blue-green strategy should cause no outage
- store terraform state in an Azure storage account

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally
- [ ] Product review

### CODE REVIEW

**Comment by karthikdfe** [comment]
[Deploy find (ftt) to PAAS with Terraform](https://trello.com/c/LDkN1Bg8/3366-deploy-find-ftt-to-paas-with-terraform)

File: azure-pipelines.yml [hunk 1]
```diff
@@ -108,7 +108,7 @@ steps:
   inputs:
     Contents: |
      azure/**
-     terraform/**
+     terraform/statuscake/**
```
Comment (saliceti) [inline azure-pipelines.yml:111]: Can this be split to a separate commit?

File: terraform/paas/backend.tf [hunk 2]
```diff
@@ -0,0 +1,5 @@
+terraform {
+  backend "azurerm" {
+    container_name       = "paas-tfstate"
+  }
```
Comment (saliceti) [inline terraform/paas/backend.tf:4]: Can we add the state file key here?
`paas-find.tfstate` for example

File: terraform/paas/data.tf [hunk 3]
```diff
@@ -0,0 +1,12 @@
+data "cloudfoundry_domain" "local" {
+  name = "london.cloudapps.digital"
+}
+
+data "cloudfoundry_org" "org" {
+  name = "dfe-teacher-services"
+}
+
+data "cloudfoundry_space" "space" {
+  name = "find-qa"
```
Comment (saliceti) [inline terraform/paas/data.tf:10]: This should come form a variable

File: terraform/paas/resources.tf [hunk 4]
```diff
@@ -0,0 +1,104 @@
+resource "cloudfoundry_app" "find-teacher-training" {
+  for_each =  var.app
```
Comment (saliceti) [inline terraform/paas/resources.tf:2]: As discussed, we will only deploy one app (find) to one environment in each terraform run. Therefore we can simplify and avoid the for_each loop.

File: terraform/paas/resources.tf [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    SECRET_KEY_BASE                                  = var.SECRET_KEY_BASE
+    SENTRY_DSN                                       = var.SENTRY_DSN
+    SETTINGS__GOOGLE__GCP_API_KEY                    = var.SETTINGS__GOOGLE__GCP_API_KEY
+    SETTINGS__GOOGLE__MAPS_API_KEY                   = var.SETTINGS__GOOGLE__MAPS_API_KEY
+    SETTINGS__REDIRECT_RESULTS_TO_C_SHARP            = each.value.environment.SETTINGS__REDIRECT_RESULTS_TO_C_SHARP
+    WEBPACKER_DEV_SERVER_HOST                        = each.value.environment.WEBPACKER_DEV_SERVER_HOST
+    WEBSITE_SLOT_POLL_WORKER_FOR_CHANGE_NOTIFICATION = each.value.environment.WEBSITE_SLOT_POLL_WORKER_FOR_CHANGE_NOTIFICATION
+  }
+  timeout = 900
+  ports   = [3000]
```
Comment (saliceti) [inline terraform/paas/resources.tf:21]: Can we double check? It's not [in the doc](https://github.com/cloudfoundry-community/terraform-provider-cf/wiki/resource_app)

File: terraform/paas/resources.tf [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    RAILS_SERVE_STATIC_FILES                         = each.value.environment.RAILS_SERVE_STATIC_FILES
+    SECRET_KEY_BASE                                  = var.SECRET_KEY_BASE
+    SENTRY_DSN                                       = var.SENTRY_DSN
+    SETTINGS__GOOGLE__GCP_API_KEY                    = var.SETTINGS__GOOGLE__GCP_API_KEY
+    SETTINGS__GOOGLE__MAPS_API_KEY                   = var.SETTINGS__GOOGLE__MAPS_API_KEY
+    SETTINGS__REDIRECT_RESULTS_TO_C_SHARP            = each.value.environment.SETTINGS__REDIRECT_RESULTS_TO_C_SHARP
+    WEBPACKER_DEV_SERVER_HOST                        = each.value.environment.WEBPACKER_DEV_SERVER_HOST
+    WEBSITE_SLOT_POLL_WORKER_FOR_CHANGE_NOTIFICATION = each.value.environment.WEBSITE_SLOT_POLL_WORKER_FOR_CHANGE_NOTIFICATION
+  }
+  timeout = 900
```
Comment (saliceti) [inline terraform/paas/resources.tf:20]: Shall we try to remove this timeout and check if there is an issue with startup time?

File: terraform/paas/resources.tf [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+  ports   = [3000]
+}
```
Comment (saliceti) [inline terraform/paas/resources.tf:22]: Can we try the [native blue green deploy strategy](https://github.com/cloudfoundry-community/terraform-provider-cf/wiki/resource_app#application-deployment-strategy-blue-green-deploy)?

File: terraform/paas/backend.tf [hunk 8]
```diff
@@ -0,0 +1,6 @@
+terraform {
+  backend "azurerm" {
+    container_name = "paas-tfstate"
+    key            = "paas-qa-find.tfstate"
```
Comment (saliceti) [inline terraform/paas/backend.tf:4]: This can't have hard coded environment here as it is the same code for multiple environments.
I suggest to use just "paas-find.tfstate"

File: terraform/paas/terraform_qa.tfvars [hunk 9]
```diff
@@ -0,0 +1,15 @@
+app = {
+  name         = "find-postgraduate-teacher-training"
+  docker_image = "dfedigital/find-teacher-training:latest"
+  hostname     = "qa-find-postgraduate-teacher-training"
+  space        = "find-qa"
+}
+
+env = {
```
Comment (saliceti) [inline terraform/paas/terraform_qa.tfvars:8]: "env" may be confusing. Can we rename it to "app_env"?

**Comment by saliceti** [comment]
Looks like the statuscake terraform code was deleted? Or was it moved?
